### PR TITLE
[T5A2][T13-B1] Low Bao Ling Vivian

### PR DESCRIPTION
**Pull request by sunset1215**

### CODE REVIEW

File: src/seedu/addressbook/commands/Command.java [hunk 1]
```diff
@@ -38,8 +38,14 @@ public static String getMessageForPersonListShownSummary(List<? extends ReadOnly
 
     /**
      * Executes the command and returns the result.
+     * @throws Exception 
      */
-    public abstract CommandResult execute();
+    public abstract CommandResult execute() throws Exception;
+    
+    /**
+     * Returns true if command changes the data
```
Comment (okkhoy) [inline src/seedu/addressbook/commands/Command.java]: try to be more specific on this header comment; the abstract class is a contract that we are providing to the client classes, hence the documentation should be precise. e.g., think what happens when you try to add a duplicate entry to address book. does it still mutate data?

File: src/seedu/addressbook/logic/Logic.java [hunk 2]
```diff
@@ -85,7 +85,9 @@ public CommandResult execute(String userCommandText) throws Exception {
     private CommandResult execute(Command command) throws Exception {
         command.setData(addressBook, lastShownList);
         CommandResult result = command.execute();
-        storage.save(addressBook);
+        if (command.isMutating()) {
+            storage.save(addressBook);
```
Comment (okkhoy) [inline src/seedu/addressbook/logic/Logic.java]: may be you need to check how `LogicTest` works now? does it fail?

**Comment by okkhoy** [comment]
@sunset1215 
Some comments added. Please ack & close the PR after reading comments.

**Comment by sunset1215** [comment]
@okkhoy 
I added a boolean variable isMutating (default value = false) for AddCommand and DeleteCommand so that both commands are mutating only when the add or delete operation was executed successfully.

The LogicTest still fails, but I couldn't find the reason why. When I tried debugging by inserting temporary print statements, the logic appears to work. After examining the LogicTest further, I found a discrepancy in the method assertCommandBehavior (maybe...).

codesnip

Line 72 states that storage file content should match expectedAddressBook. I presume this corresponds to Line 93, but it's comparing the storage file content with the internal addressBook instead. I thought this might be an error, so I tried changing it to expectedAddressBook. But the test case still fails. Closing this PR now because I have to work on other modules, but would really appreciate a hint or explanation if possible. Thanks!!

**Comment by okkhoy** [comment]
Yeah, so here goes.
- the last `assertEquals` (line 93) reads from the file, rebuilds the object and does the comparison.
- until now, all commands were saving to the file, hence this test was passing.
why?
 1. because, the test `setup` does a default saving (check line 39 in `LogicTest`) (which is an empty file). 
 2. when the command does a save, it rewrites the contents of the default saved empty file 
 3. when you reload in the test, you get the saved contents.
 4. with the changed behavior, the comparison is happening against the empty file that was saved in the `setup`. Hence the tests fail.
how to solve?
- one of the ways is to remove the last assert in `assertCommandBehavior`  for all commands.
- and have a separate check for mutating commands. (this is necessary due to the new behavior introduced)
I have started a discussion with Prof on this, if he suggests any alternative, I will relay the information.

**Comment by sunset1215** [comment]
Ah, I think I get it now. Thanks a lot!!

**Pull request closed without merging**